### PR TITLE
Link to current branch in More's Compare and Commits links

### LINKED ISSUE #1197: "Compare" repo link should include the current branch/tag
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/refined-github/issues/1197)

1. Visit https://github.com/BrowserSync/browser-sync/tree/client-ts
2. Open repo tab "More" -> "Compare"
3. See a plain `master` -> `master` comparison: https://github.com/BrowserSync/browser-sync/compare

I'd expect that to include a reference to the current branch/commit/tag, like:

https://github.com/BrowserSync/browser-sync/compare/master...client-ts

Maybe there's a better way, but I'm trying to see what changed between `master` and whatever branch I'm looking at.

This suggestion is similar to #966, but not just between tags.

The default branch ~~can be determined via fetch, exactly like in #1173~~ is available via https://github.com/sindresorhus/refined-github/blob/master/source/libs/get-default-branch.js

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Sponsors (Total: $40.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/refined-github/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### PR DESCRIPTION
**Pull request by jerone**
### Description
Include current ref (branch or tag) in compare & commit repo links added in [`more-dropdown`](https://github.com/sindresorhus/refined-github/blob/master/source/features/more-dropdown.tsx#L37-L48). Different pages required different detection to get the current ref (branch or tag).

### Extra
Optionally I can include [the default branch](https://github.com/sindresorhus/refined-github/blob/master/source/libs/get-default-branch.ts), but that requires an extra API call and is not necessary.

### Closes
Closes #1197.

### Test
1. Visit any of the below urls.
2. Open repo tab "More" -> "Compare" or "Commits".
3. See it link to the correct tree (branch or tag).

* https://github.com/BrowserSync/browser-sync ➡ no ref
* https://github.com/BrowserSync/browser-sync/tree/client-ts ➡ compare & commits should link to ref
* https://github.com/BrowserSync/browser-sync/compare/client-ts ➡ compare & commits should link to ref
* https://github.com/BrowserSync/browser-sync/compare/62007c8b944808d1b46d42d5e22fa65883d1eaec ➡ compare & commits should link to ref
* https://github.com/BrowserSync/browser-sync/compare/client-ts...master ➡ compare & commits should link to the left ref.
* https://github.com/BrowserSync/browser-sync/commits/client-ts ➡ compare & commits should link to ref
* https://github.com/BrowserSync/browser-sync/commits/62007c8b944808d1b46d42d5e22fa65883d1eaec ➡ compare & commits should link to ref
* https://github.com/BrowserSync/browser-sync/releases/tag/v2.26.3 ➡ compare & commits should link to ref
* https://github.com/BrowserSync/browser-sync/compare ➡ no ref
* https://github.com/BrowserSync/browser-sync/commits ➡ will default and link to master
* https://github.com/BrowserSync/browser-sync/blob/master/README.md  ➡ compare & commits should link to ref
* https://github.com/BrowserSync/browser-sync/wiki/Release-history ➡ no ref; should not link to "Release-history"

### CODE REVIEW

File: source/features/more-dropdown.tsx [hunk 1]
```diff
@@ -31,10 +31,32 @@ async function init(): Promise<void> {
 		createDropdown();
 	}
 
+	let compareUrl = `/${repoUrl}/compare`;
+	let commitsUrl = `/${repoUrl}/commits`;
+	let tree = '';
```
Comment (bfred-it) [inline source/features/more-dropdown.tsx]:
```suggestion
	let ref = '';
```

I think `ref` is generic `git` term for this.

File: source/features/more-dropdown.tsx [hunk 2]
```diff
@@ -31,10 +31,32 @@ async function init(): Promise<void> {
...
+		if (urlParts[4].includes('...')) {
+			tree = urlParts[4].split('...')[0];
+		} else {
+			tree = urlParts[4];
+		}
+	} else if (urlParts[3] === 'commits') { // On commits page
+		tree = urlParts[4];
+	} else if (urlParts[3] === 'releases' && urlParts[4] === 'tag') { // On tag page
+		tree = urlParts[5];
+	}
```
Comment (bfred-it) [inline source/features/more-dropdown.tsx]: Maybe:

```js
if (isCompare() && location.pathname.includes('...')) {
	[ref] = pathnameParts[4].split('...');
} else {
	ref = pathnameParts[isTagsOrReleases() ? 5 : 4]
}
```

File: source/features/more-dropdown.tsx [hunk 3]
```diff
@@ -31,10 +31,32 @@ async function init(): Promise<void> {
 		createDropdown();
 	}
 
+	let compareUrl = `/${repoUrl}/compare`;
+	let commitsUrl = `/${repoUrl}/commits`;
+	let tree = '';
+	const urlParts = location.pathname.split('/');
```
Comment (bfred-it) [inline source/features/more-dropdown.tsx]:
```suggestion
	const pathnameParts = location.pathname.split('/');
```

File: source/features/more-dropdown.tsx [hunk 4]
```diff
@@ -31,10 +31,32 @@ async function init(): Promise<void> {
 		createDropdown();
 	}
 
+	let compareUrl = `/${repoUrl}/compare`;
+	let commitsUrl = `/${repoUrl}/commits`;
```
Comment (bfred-it) [inline source/features/more-dropdown.tsx]: These two can be before the `if tree` below

**Comment by jerone** [comment]
Processed your feedback and fixed linting.

File: source/features/more-dropdown.tsx (hunk 2 above)
Comment (bfred-it) [reply source/features/more-dropdown.tsx]: Sorry, I think this broke it, not all pages have a ref. See if this works

```js
if (isCompare()) {
	[ref] = pathnameParts[4].split('...');
} else if (isTagsOrReleases()) {
	ref = pathnameParts[5]
} else if (['commits', 'blob', 'tree'].includes(pathnameParts[3])) {
	ref = pathnameParts[4]
}
```

Comment (jerone) [reply source/features/more-dropdown.tsx]: Fixed compare detection.
Added blame detection.
Prioritized order.

And moved ref detection to it's own function in utils.js. And also included tests.

**Comment by bfred-it** [comment]
I don't think this should affect Compare pages. The only reason why I'd use that link when already on the Compare pages would be to "reset the view"

Tags shouldn't be affected either, e.g. https://github.com/BrowserSync/browser-sync/releases/tag/v2.26.3, for two reasons:
- The generated link compares to master and it's always empty (`master...v2.26.3`)
- There's already a compare link on the page named "8 commits on master since this tag", which compares in the right direction  (`v2.26.3...master`)

**Comment by jerone** [comment]
Removed compare & tags/releases detection.

**Comment by bfred-it** [comment]
I just saw isPRCommit too, that also should probably be dropped, it's too specific I think

**Comment by jerone** [comment]
Removed PR commit detection too.

**Comment by bfred-it** [comment]
Thanks @jerone 🍫

**Comment by jerone** [comment]
More LOC for tests, then for the real fix. 🎉😂

**Pull request merged**